exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,407 @@
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(clusterResourceSet, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to Patch the ClusterResourceSet object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, clusterResourceSet, patch.WithStatusObservedGeneration{}); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	logger := r.Log.WithValues("clusterresourceset", clusterResourceSet.Name, "namespace", clusterResourceSet.Namespace)
+
+	clusters, err := r.getClustersByClusterResourceSetSelector(ctx, clusterResourceSet)
+	if err != nil {
+		logger.Error(err, "Failed fetching clusters that matches ClusterResourceSet labels", "ClusterResourceSet", clusterResourceSet.Name)
+		conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.ClusterMatchFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	for _, cluster := range clusters {
+		if err := r.ApplyClusterResourceSet(ctx, cluster, clusterResourceSet); err != nil {
+			// The reason of not requeuing in case of errors if applying resources are failed is to avoid retries in case resources are missing.
+			// In the next reconcile, failed resources will be retried.
+			logger.Error(err, "Failed applying resources to cluster", "Cluster", cluster.Name)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// getClustersByClusterResourceSetSelector fetches Clusters matched by the ClusterResourceSet's label selector that are in the same namespace as the ClusterResourceSet object.
+func (r *ClusterResourceSetReconciler) getClustersByClusterResourceSetSelector(ctx context.Context, clusterResourceSet *addonsv1.ClusterResourceSet) ([]*clusterv1.Cluster, error) {
+	logger := r.Log.WithValues("clusterresourceset", clusterResourceSet.Name, "namespace", clusterResourceSet.Namespace)
+
+	clusterList := &clusterv1.ClusterList{}
+	selector, err := metav1.LabelSelectorAsSelector(&clusterResourceSet.Spec.ClusterSelector)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to convert selector")
+	}
+
+	// If a ClusterResourceSet has a nil or empty selector, it should match nothing, not everything.
+	if selector.Empty() {
+		logger.Info("Empty ClusterResourceSet selector: No clusters are selected.")
+		return nil, nil
+	}
+
+	if err := r.Client.List(ctx, clusterList, client.InNamespace(clusterResourceSet.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		return nil, errors.Wrap(err, "failed to list clusters")
+	}
+
+	clusters := []*clusterv1.Cluster{}
+	for i := range clusterList.Items {
+		c := &clusterList.Items[i]
+		if c.DeletionTimestamp.IsZero() {
+			clusters = append(clusters, c)
+		}
+	}
+	return clusters, nil
+}
+
+// ApplyClusterResourceSet applies resources in a ClusterResourceSet to a Cluster. Once applied, a record will be added to the
+// cluster's ClusterResourceSetBinding.
```

Comment:
Am I correct in saying that Binding objects are used only internally and don't have a reconciler associated with it?

Reply:
It doesn't have a reconciler. We have a Binding object per cluster for showing the status of all CRSs that matched with that cluster.